afs-cassandra/src/main/java/com/powsybl/afs/cassandra/CassandraAppStorage.java

```diff
@@ -995,9 +996,25 @@ public void write(int b) throws IOException {
 
         @Override
         public void write(byte[] b, int off, int len) throws IOException {
-            gzos.write(b, off, len);
-            count += len;
-            executeIfNecessary();
+            if (len > config.getBinaryDataChunkSize()) {
+                int chunkOffset = off;
+                int chunkCount = 0;
+                while (chunkCount * config.getBinaryDataChunkSize() <= len) {
+                    int chunkLen = config.getBinaryDataChunkSize();
+                    if (len - chunkOffset < config.getBinaryDataChunkSize()) {
+                        chunkLen = len - chunkOffset;
+                    }
+                    gzos.write(Arrays.copyOfRange(b, chunkOffset, chunkOffset + chunkLen));
+                    count += chunkLen;
+                    executeIfNecessary();
+                    chunkCount += 1;
+                    chunkOffset = off + chunkCount * config.getBinaryDataChunkSize();
+                }
+            } else {
+                gzos.write(b, off, len);
+                count += len;
+                executeIfNecessary();
+            }
```

Comment:
Seems not OK to me: if you start writing X bytes (x lower than the chunk size), what is expected for the second write?
From my point of view we should complete the current chunk, write several full chunks, and start a new chunk with the rest of the data. For instance, with a `chunkSize = 10`:
```
write("aaaaa", 0, 5);
// At this stage, the current chunk contains "aaaaa", count is equal to 5 and chunkCount is equal to 0

write("bbbbbbbbbb", 0, 10); 
// At this stage, the first chunk has been completed by the 5 first B characters.
// As the count is equal to the chunkSize, execute() is called
// A new chunk is created with the 5 last "bbbbb"

// From here, there are two options
// Continue to write data or close the ouputStream that will flush the chunk

Reply:
I agree with that analysis.

I think we should have more isolated unit tests for those input/output streams, maybe re-writing this class as `ChunkedOutputStream`, which would not necessarily have Cassandra as underlying storage, for example.

Also, while reading again the code for the read part, I wonder if there are not problems on that side also (when calling read with a length larger than the chunk size, for example).

Reply:
Indeed, I didn't covered the case when len is variable (with preceding length write being under the chunk size and after greater).
Anyway, as stated by @sylvlecl , the resulting chunk data size written in db is not the one enforced for input data, and could be variable depending on the gz compression. So if we really want to enforce a strict size for chunks (except the last one), we should make a deeper refactoring with the GzipOutputStream..
But do we really want to enforce that ?

Though wrongly titled, this PR prevent a data chunk to be written with a size greater than a defined max size, in order (I guess) to prevent reaching the cassandra storage limits (or any other db for that matter) :  https://cwiki.apache.org/confluence/display/CASSANDRA2/CassandraLimitations)
Is there really a need beyond that, to enforce the strict size of chunks (that would require to handle the size of the compressed content instead) ?

Reply:
Agreed that this is enough as a fix.
We can have a deeper/cleaner refactoring in a later evolution.

Reply:
Well, the code was not correct afterall, so I fixed it and added the proper code to fill in previous chunks writes (even if at the end the real byte size is not correct due to compression).

Reply:
@pl-buiquang @sylvlecl Could you write an issue to explain a little bit the refactoring we expect for this part so we won't forget it.
As I mention in my previous comment, I don't think this is a good idea to enforce the compressed chunk be equal to the chunk length. We can expect the size of compressed data is smaller than this limits, but it's not always true. For instance, if you compress an empty string, the size of the compressed data is larger that the size of uncompressed data. It could also be the case if the data contains only different characters, and no redundant patterns

Reply:
For information, I have mentioned the ideas of refactoring and of enforcing the size limitations on stored chunks in that issue #59 .